### PR TITLE
Add E0128, E0130, E0134, E0135 and E0159 error explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #24407.
r? @Manishearth

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -1368,6 +1368,52 @@ struct Foo {
...
+You declared a type parameter with a default which has the same identifier as
+the type parameter. Erroneous code example:
+
+```
+pub struct Foo<SomeType=SomeType>;
+// error: type parameters with a default cannot use forward declared
+// identifiers
+```
+
+Please verify you used the good type or change the type parameter identifier.
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: "Please verify that a name clash hasn't been accidentally introduced, and rename the type parameter if so"

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -1368,6 +1368,52 @@ struct Foo {
...
+Erroneous code example:
+
+```
+extern {
+    fn foo((a, b): (u32, u32)); // error: patterns aren't allowed in foreign
+                                //        function declarations
+}
+```
+
+Please remove the pattern argument in the foreign function declaration and
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: Not necessary to define  a struct. We're just declaring it, so the pattern can be replaced by a direct binding, e.g. `x: (u32, u32)`

"Please replace the pattern argument with a regular one"

File: src/librustc/diagnostics.rs [hunk 3]
```diff
@@ -411,6 +411,42 @@ fn main() {
...
+You tried to modify the str type, which isn't allowed. Erroneous code
+example:
+
+```
+let s = "salut";
+let c = &mut (*s)[0..1]; // error: modification of string types is not
+                         //        allowed
+```
+
+I you want to modify an str, please use the String type. Example:
```
Comment (apasel422) [inline src/librustc/diagnostics.rs]: s/I/If/

File: src/librustc/diagnostics.rs [hunk 4]
```diff
@@ -411,6 +411,42 @@ fn main() {
 See also https://doc.rust-lang.org/book/unsafe.html
 "##,
 
+E0134: r##"
```
Comment (arielb1) [inline src/librustc/diagnostics.rs]: Aren't these error-codes borrowck's business?

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -1368,6 +1368,62 @@ struct Foo {
...
+Since type parameters are evaluated in-order, you can fix this issue by doing:
+
+```
+pub struct Foo<U=(), T=U> {
+    field1: T,
+    filed2: U,
+}
+```
+
+Please also verify that a name clash hasn't been accidentally introduced, and
```
Comment (arielb1) [inline src/librustc_typeck/diagnostics.rs]: Nit: name-clashes can't be introduced by reordering. Maybe "that this wasn't because of a name-clash".

**Pull request merged**